Rework key types management and update new master key in the stash

jrisc opened the pull request:
## Key types management

### Problem

Standalone MIT KDC instances are using the `supported_enctypes` `kdc.conf` parameter to determine which key types are accepted when setting a principal's credentials. The order in which keys are saved in the KDB is the same as the types provided by the admin. If the admin did not provide an explicit list of key types, then the full list of `supported_enctypes` will be used (still preserving the ordering).

The ordering of principal keys is important because it determines which encryption type will be use the generate the ticket to this principal. When the KDC processes a ticket request, the session key type is selected based on the list provided by the client (depending on its `permitted_enctypes`, `default_tgs_enctypes`, or `default_tkt_enctypes` krb5 configuration parameter), and the KDC's `permitted_enctypes`. But the ticket encryption type itself will be the one from the first principal key allowed by the KDC's `permitted_enctypes`. Here `permitted_enctypes` only acts as a filter, it has no impact on encryption type priority.

As a consequence, the explicit list of key types provided by the admin or the list of default key types (i.e. `supported_entypes`) at the time the principal credentials are set, will impact the encryption type used for generating the ticket to this principal until the credentials are modified again.
    
IPA separates the notions of default and supported key types. So far, they were configured as the LDAP attributes `krbDefaultEncSaltTypes` and `krbSupportedEncSaltTypes` respectively. However some design mistakes in the way these attributes where stored and updated may cause key types to be read out-of-order (i.e. not the strongest type first and the weakest last). But as explained earlier, fixing these attributes alone will not fix the keys that were generated in the meantime, and might already be stored out-of-order.

### Solution

Hence this PR introduces 2 changes to fix this problem:

* The `krbDefaultEncSaltTypes` and `krbSupportedEncSaltTypes` LDAP attributes are now ignored. Instead, the default key types, supported key types, and initial master key type are inferred from the IPA server's `permitted_enctypes` in the krb5 configuration. This will ensure there is no mismatch between the local and domain-level configurations. It will also make IPA more portable because it removes hard-coded encryption types settings, to function according to its runtime environment instead.

* When a credential update request features an explicit key types list, it will systematically be reordered and filtered according to the server's `permitted_enctypes` setting. And so it is for existing list of keys in principal entries.

This allows to have almost all Kerberos key operations relying on the value of the `permitted_enctypes` configuration parameter. The only exception is randomized key generation using kadmin, which does not go through the KDB DAL. Hence it still relies on `supported_enctypes` from `kdc.conf`. Solving this will require a MIT krb5 upstream change.

However, this PR include a warning that will be raised at KDC starting if a difference between `permitted_enctypes` and `supported_enctypes` is detected.

### Additional changes

To be mentioned that putting this change in place required to modify the output format of `ipa-getkeytab --permitted-enctypes`, which is now using the same format of encryption type names used in krb5 configuration files. This command is now used as part of a domain's initial setup.

This PR also fixes the way the IPA SLAPI extop-pwd plugin deals with default key/salt types list: special salt type was used systematically, while this is irrelevant for randomized keys. Now, when no explicit key/salt types list is provided by the user, the permitted encryption types are used in conjunction with normal salt type for randomized key, or special salt type for password.

Similar modifications are applied in `ipa-getkeytab` to for using the right salt types when keys are generated on the client side.

## Master key upgrade

* Documentation of master key manual upgrade
* Inclusion of test_mkey_upgrade in nightly tests

## Update krb5 dependencies

The recent IPA commit 43b1fd77f10cf2752a44b4b5c219660872e5b1de requires the MIT krb5 commit `4ed7da378940198cf4415f86d4eb013de6ac6455` for the master key switch to actually take effect. This krb5 fix was already backported to Fedora and RHEL.

Fixes: https://pagure.io/freeipa/issue/9370
Fixes: https://pagure.io/freeipa/issue/9840

jrisc committed:
freeipa.spec.in: update krb5 dependencies

The recent IPA commit 43b1fd77f10cf2752a44b4b5c219660872e5b1de requires
the MIT krb5 commit 4ed7da378940198cf4415f86d4eb013de6ac6455 for the
master key switch to actually take effect. This krb5 fix was already
backported to Fedora and RHEL.

Fixes: https://pagure.io/freeipa/issue/9370
Signed-off-by: Julien Rische <jrische@redhat.com>

flo-renaud commented:
Hi @jrisc 
you need to rebase on top of master branch in order to pick the commit https://pagure.io/freeipa/c/2eaba8497a5095b23dac39b759dbf632fa422529?branch=master that fixes test_adtrust_install.

jrisc committed:
doc/designs: add manual procedure for master key upgrade

Signed-off-by: Julien Rische <jrische@redhat.com>

flo-renaud commented on doc/designs/master-key-upgrade.md:
````diff
@@ -0,0 +1,100 @@
+Re-enter KDC database master key to verify: 
+```
+
+The new key should now appear in the master key list and have the `aes256-cts-hmac-sha384-192` encryption type. The old key should have a `*` character at the end of its line. This means it is still the active master key, which means new key are still going to be encrypted using the old master key at this point
````
typo: new key are -> new keys are

flo-renaud commented on doc/designs/master-key-upgrade.md:
````diff
@@ -0,0 +1,100 @@
+KVNO: 1, Enctype: aes256-cts-hmac-sha1-96, Active on: Wed Dec 31 19:00:00 EST 1969 *
+```
+
+The next still is set the new master key as the active one. To do so, use the hereunder command with the KVNO of the new master key (in this example, the KVNO is 2):
````
Typo: still is set -> step is to set

flo-renaud commented on doc/designs/master-key-upgrade.md:
````diff
@@ -0,0 +1,100 @@
+# kdb5_util use_mkey 2
+```
+
+Both master keys should still be there afterwards, but now the new key has the `*` character. It signify it is the new active key:
````
typo: signify -> signifies

flo-renaud commented on doc/designs/master-key-upgrade.md:
````diff
@@ -0,0 +1,100 @@
+KVNO: 1, Enctype: aes256-cts-hmac-sha1-96, Active on: Wed Dec 31 19:00:00 EST 1969
+```
+
+So at this point, the new keys will be encrypted using the new master key, but the already existing key are still using the master key. If you query a principal whose credentials were not updated since the new was set as active, it should still refer to the old master key (here the master key with KVNO 1):
````
but the already existing **keys** are still using the **old** master key

flo-renaud commented on doc/designs/master-key-upgrade.md:
````diff
@@ -0,0 +1,100 @@
+MKey: vno 1
+```
+
+This is why, before removing the old master key, all the other keys in the database have to be re-encrypted using the new active master key. This operation will send a lot of modification requests to the database, so it preferable to execute it on a moment when the FreeIPA replicas are the least busy. Also be sure to keep the `-x unlockiter` option for not locking the database during the process:
````
so it preferable -> it **is** preferable

flo-renaud reviewed (commented):
Hi @jrisc 

please find inline comments

flo-renaud reviewed (commented):
Hi @jrisc 

Thanks for the updated doc. I have a few inline comments + the following ones:
- the doc should also mention the other replicas. I assume there is no need to perform anything on the other servers as the key will be replicated, but there may be a need to restart IPA services.
- what about the existing user/host/service keys? The doc should explain that they remain unchanged and there is no need to regenerate keytab files (the old keytabs should still be working).
- a new user created after the migration is showing MKey: vno1. Is it expected?
```
# kadmin.local getprinc testuser2@IPA.TEST
Principal: testuser2@IPA.TEST
Expiration date: [never]
Last password change: Thu Jul 03 05:53:53 EDT 2025
Password expiration date: Thu Jul 03 05:53:53 EDT 2025
Maximum ticket life: 1 day 00:00:00
Maximum renewable life: 7 days 00:00:00
Last modified: Thu Jul 03 05:53:53 EDT 2025 (root/admin@IPA.TEST)
Last successful authentication: [never]
Last failed authentication: [never]
Failed password attempts: 0
Number of keys: 2
Key: vno 1, aes256-cts-hmac-sha1-96:special
Key: vno 1, aes128-cts-hmac-sha1-96:special
MKey: vno 1
Attributes: REQUIRES_PRE_AUTH DISALLOW_SVR
Policy: [none]
```
- after the master key has been migrated, ipa-getkeytab command is failing:
```
# ipa-getkeytab -v -s `hostname` -p host/`hostname`@IPA.TEST -D "cn=directory manager" -w Secret123 -r -k /etc/krb5.keytab
Using provided server replica.ipa.test
Failed to parse result: Internal Error.
Failed to get keytab
```
and we see the following error in 389-ds error log: `ERR - encode_getkeytab_reply - [file ipa_pwd_extop.c, line 1497]: Failed to decrypt keys`
Is there any additional step missing in the migration procedure?

jrisc commented:
I have rerun the process, and it seems indeed there is a problem with the reference to the master key: after restarting the IPA server, the existing principals are working, but the newly create ones are still referencing the deleted old master key.

I suspect there is a bug in ipa-kdb which causes the new entries to keep referencing master key version 1, regardless of the active master key.

jrisc committed:
ipa-kdb: replace old mkey by new one in the stash

When executing `kdb5_util add_mkey`, ipa-kdb was adding a "krbMKey"
attribute in 389ds, the old "krbMKey" was not removed. However,
ipa-pwd-extop supports only one master key, which was causing it to only
consider the first attribute (i.e. the old master key).

This commit fixes this problem by having ipa-kdb completely replacing
the old krbMKey attribute by the new one.

Signed-off-by: Julien Rische <jrische@redhat.com>

jrisc committed:
ipa-kdb: fix a potential memory leak

Signed-off-by: Julien Rische <jrische@redhat.com>

jrisc commented:
The dedicated test [fedora-latest/test_mkey_upgrade](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/afe78268-73b3-11f0-a789-fa163e2c4571) is passing.

The 5 failures left are known issues:
* [fedora-latest/test_fips](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/a84774ec-76b9-11f0-9fc1-fa163ec236b6): https://pagure.io/freeipa/issue/9785
* [fedora-latest/test_installation_TestInstallWithCA_DNS3](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/b0120e12-76b9-11f0-af2b-fa163ef026f5): https://pagure.io/freeipa/issue/9135
* [fedora-latest/test_installation_TestKRAinstallAfterCertRenew](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/bffbe050-76b9-11f0-8eab-fa163e57b661): https://pagure.io/freeipa/issue/9763
* [fedora-latest/test_ipahealthcheck_cli_fsspace](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/c7aacdac-76be-11f0-b8f9-fa163e396e38): https://github.com/dogtagpki/pki/issues/4906
* [fedora-latest/test_topology_TestCASpecificRUVs](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/c40c1f52-76b9-11f0-a9fa-fa163ef83758): https://pagure.io/freeipa/issue/8023

flo-renaud commented:
Manual testing detected an issue in the part upgrading the supported encryption types. After upgrade, we have:
```
dn: cn=IPA.TEST,cn=kerberos,dc=ipa,dc=test
krbMKey:: MCoCAQIwJQIBFAQgtKQ307Zh5HhaJlEUQdqPphGEt7GJkC9X3YAarJtBOas=
cn: IPA.TEST
objectClass: top
objectClass: krbrealmcontainer
objectClass: krbticketpolicyaux
krbSubTrees: dc=ipa,dc=test
krbSearchScope: 2
krbSupportedEncSaltTypes: aes256-cts:normal
krbSupportedEncSaltTypes: aes256-cts:special
krbSupportedEncSaltTypes: aes128-cts:normal
krbSupportedEncSaltTypes: aes128-cts:special
krbSupportedEncSaltTypes: aes128-sha2:normal
krbSupportedEncSaltTypes: aes128-sha2:special
krbSupportedEncSaltTypes: aes256-sha2:normal
krbSupportedEncSaltTypes: aes256-sha2:special
krbSupportedEncSaltTypes:: YWVzMjU2LXNoYTI6c3BlY2lhbAphZGQ6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogYWVzMTI4LXNoYTI6c3BlY2lhbAphZGQ6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogYWVzMjU2LXNoYTI6bm9ybWFsCmFkZDoga3JiU3VwcG9ydGVkRW5jU2FsdFR5cGVzOiBhZXMxMjgtc2hhMjpub3JtYWwKYWRkOiBrcmJTdXBwb3J0ZWRFbmNTYWx0VHlwZXM6IGFlczI1Ni1jdHM6c3BlY2lhbAphZGQ6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogYWVzMTI4LWN0czpzcGVjaWFsCmFkZDoga3JiU3VwcG9ydGVkRW5jU2FsdFR5cGVzOiBhZXMyNTYtY3RzOm5vcm1hbAphZGQ6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogYWVzMTI4LWN0czpub3JtYWwKYWRkOiBrcmJEZWZhdWx0RW5jU2FsdFR5cGVzOiBhZXMyNTYtc2hhMjpzcGVjaWFsCmFkZDoga3JiRGVmYXVsdEVuY1NhbHRUeXBlczogYWVzMTI4LXNoYTI6c3BlY2lhbAphZGQ6IGtyYkRlZmF1bHRFbmNTYWx0VHlwZXM6IGFlczI1Ni1jdHM6c3BlY2lhbAphZGQ6IGtyYkRlZmF1bHRFbmNTYWx0VHlwZXM6IGFlczEyOC1jdHM6c3BlY2lhbApyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogZApyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogZQpyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogcwpyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogMwpyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogLQpyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogaApyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogbQpyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogYQpyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogYwpyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogLQpyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogcwpyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogaApyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogYQpyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogMQpyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogOgpyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogcwpyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogcApyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogZQpyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogYwpyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogaQpyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogYQpyZW1vdmU6IGtyYlN1cHBvcnRlZEVuY1NhbHRUeXBlczogbApyZW1vdmU6IGtyYkRlZmF1bHRFbmNTYWx0VHlwZXM6IGRlczMtaG1hYy1zaGExOnNwZWNpYWwKcmVtb3ZlOiBrcmJEZWZhdWx0RW5jU2FsdFR5cGVzOiBhcmNmb3VyLWhtYWM6c3BlY2lhbA==
krbMaxTicketLife: 86400
krbMaxRenewableAge: 604800
krbDefaultEncSaltTypes: aes256-cts:special
krbDefaultEncSaltTypes: aes128-cts:special
krbPwdPolicyReference: cn=Default Kerberos Service Password Policy,cn=Kerberos Service Password Policy,cn=IPA.TEST,cn=kerberos,dc=ipa,dc=test
```

and the decoding of the blob gives:
```
aes256-sha2:special
add: krbSupportedEncSaltTypes: aes128-sha2:special
add: krbSupportedEncSaltTypes: aes256-sha2:normal
add: krbSupportedEncSaltTypes: aes128-sha2:normal
add: krbSupportedEncSaltTypes: aes256-cts:special
add: krbSupportedEncSaltTypes: aes128-cts:special
add: krbSupportedEncSaltTypes: aes256-cts:normal
add: krbSupportedEncSaltTypes: aes128-cts:normal
add: krbDefaultEncSaltTypes: aes256-sha2:special
add: krbDefaultEncSaltTypes: aes128-sha2:special
add: krbDefaultEncSaltTypes: aes256-cts:special
add: krbDefaultEncSaltTypes: aes128-cts:special
remove: krbSupportedEncSaltTypes: d
remove: krbSupportedEncSaltTypes: e
remove: krbSupportedEncSaltTypes: s
remove: krbSupportedEncSaltTypes: 3
remove: krbSupportedEncSaltTypes: -
remove: krbSupportedEncSaltTypes: h
remove: krbSupportedEncSaltTypes: m
remove: krbSupportedEncSaltTypes: a
remove: krbSupportedEncSaltTypes: c
remove: krbSupportedEncSaltTypes: -
remove: krbSupportedEncSaltTypes: s
remove: krbSupportedEncSaltTypes: h
remove: krbSupportedEncSaltTypes: a
remove: krbSupportedEncSaltTypes: 1
remove: krbSupportedEncSaltTypes: :
remove: krbSupportedEncSaltTypes: s
remove: krbSupportedEncSaltTypes: p
remove: krbSupportedEncSaltTypes: e
remove: krbSupportedEncSaltTypes: c
remove: krbSupportedEncSaltTypes: i
remove: krbSupportedEncSaltTypes: a
remove: krbSupportedEncSaltTypes: l
remove: krbDefaultEncSaltTypes: des3-hmac-sha1:special
remove: krbDefaultEncSaltTypes: arcfour-hmac:special
```

The first issue is that the tuple returned by `get_removed_supported_enctypes` in https://github.com/freeipa/freeipa/blob/76727f970c1c810b5cfd182734a4db260bb192bd/ipaplatform/base/tasks.py#L556-L557 contains a single value and should be defined as `(val, )` instead of `(val)`.

The second issue is that ldapupdate does not know how to handle the ldap modification containing multiple changes separated by \n in a single string.

jrisc commented:
I was able to add a FIPS RHEL 9 replica to the domain created using a RHEL 8.6 server. However I had to override the `krbSupportedEncSaltTypes` and `krbDefaultEncSaltTypes`, because the `update_krbenctypes` adds the AES-SHA2 enctypes add this end of the list, while they have to be at the beginning to be prioritized and allow the RHEL 9 replica installation to succeed.

jrisc committed:
ipa-slapi: fix conflicting definitions of LOG_ALERT macro

Rename LOG_ALERT macro into LOG_ALERT_S to avoid conflict with syslog's
LOG_ALERT.

Signed-off-by: Julien Rische <jrische@redhat.com>

jrisc committed:
Revert "Use ipaplatform tasks for krb5 enctypes"

This reverts commit fb12d9e14eafeaf036951e98e0d291db892afe2d.

jrisc committed:
Revert "Add test for master key upgrade"

This reverts commit fb36633e69e76eabcdd32195a1d3ad08604ab199.

jrisc committed:
ipatests: master key upgrade based on krb5 conf types

Signed-off-by: Julien Rische <jrische@redhat.com>

jrisc committed:
Rework Kerberos key types management

Standalone MIT KDC instances are using the "supported_enctypes" kdc.conf
parameter to determine which key types are accepted when setting a
principal's credentials. The order in which keys are saved in the KDB is
the same as the types provided by the admin. If the admin did not
provide an explicit list of key types, then the full list of
"supported_enctypes" will be used (still preserving the ordering).

The ordering of principal keys is important because it determines which
encryption type will be use the generate the ticket to this principal.
When the KDC processes a ticket request, the session key type is
selected based on the list provided by the client (depending on its
"permitted_enctypes", "default_tgs_enctypes", or "default_tkt_enctypes"
krb5 configuration parameter), and the KDC's "permitted_enctypes". But
the ticket encryption type itself will be the one from the first
principal key allowed by the KDC's "permitted_enctypes". Here
"permitted_enctypes" only acts as a filter, it has no impact on
encryption type priority.

As a consequence, the explicit list of key types provided by the admin
or the list of default key types (i.e. "supported_entypes") at the time
the principal credentials are set, will impact the encryption type used
for generating the ticket to this principal until the credentials are
modified again.

IPA separates the notions of default and supported key types. So far,
they were configured as the LDAP attributes "krbDefaultEncSaltTypes" and
"krbSupportedEncSaltTypes" respectively. However some design mistakes in
the way these attributes where stored and updated may cause key types to
be read out-of-order (i.e. not the strongest type first and the weakest
last). But as explained earlier, fixing these attributes alone will not
fix the keys that were generated in the meantime, and might already be
stored out-of-order.

Hence this commit introduces 2 changes to fix this problem:

* The "krbDefaultEncSaltTypes" and "krbSupportedEncSaltTypes" LDAP
  attributes are now ignored. Instead, the default key types, supported
  key types, and initial master key type are inferred from the IPA
  server's "permitted_enctypes" in the krb5 configuration. This will
  ensure there is no mismatch between the local and domain-level
  configurations. It will also make IPA more portable because it removes
  hard-coded encryption types settings, to function according to its
  runtime environment instead.

* When a credential update request features an explicit key types list,
  it will systematically be reordered and filtered according to the
  server's "permitted_enctypes" setting. And so it is for existing list
  of keys in principal entries.

To be mentioned that putting this change in place required to modify the
output format of "ipa-getkeytab --permitted-enctypes", which is now
using the same format of encryption type names used in krb5
configuration files. This command is now used as part of a domain's
initial setup.

This commit also fixes the way the IPA SLAPI extop-pwd plugin deals with
default key/salt types list: special salt type was used systematically,
while this is irrelevant for randomized keys. Now, when no explicit
key/salt types list is provided by the user, the permitted encryption
types are used in conjunction with normal salt type for randomized key,
or special salt type for password.

Fixes: https://pagure.io/freeipa/issue/9840
Signed-off-by: Julien Rische <jrische@redhat.com>

jrisc committed:
[tmp] Run test_mkey_upgrade and test_key_types in GitHub-CI

jrisc committed:
ipatests: add key types tests

Signed-off-by: Julien Rische <jrische@redhat.com>

flo-renaud commented:
Hi @jrisc 
can you fix the errors reported by lint and codestyle?

```
./ipatests/test_integration/test_key_types.py:23:1: E302 expected 2 blank lines, found 1
./ipatests/test_integration/test_key_types.py:47:13: E128 continuation line under-indented for visual indent
./ipatests/test_integration/test_key_types.py:55:81: E501 line too long (82 > 80 characters)
./ipatests/test_integration/test_key_types.py:56:81: E501 line too long (95 > 80 characters)
./ipatests/test_integration/test_key_types.py:57:81: E501 line too long (99 > 80 characters)
./ipatests/test_integration/test_key_types.py:70:1: E302 expected 2 blank lines, found 1
./ipatests/test_integration/test_key_types.py:72:13: E221 multiple spaces before operator
./ipatests/test_integration/test_key_types.py:74:1: E302 expected 2 blank lines, found 1
./ipatests/test_integration/test_key_types.py:76:13: E221 multiple spaces before operator
./ipatests/test_integration/test_key_types.py:78:1: E302 expected 2 blank lines, found 1
./ipatests/test_integration/test_key_types.py:80:13: E221 multiple spaces before operator
./ipatests/test_integration/test_key_types.py:82:1: E302 expected 2 blank lines, found 1
./ipatests/test_integration/test_key_types.py:84:13: E221 multiple spaces before operator
./ipatests/test_integration/test_key_types.py:86:1: E302 expected 2 blank lines, found 1
./ipatests/test_integration/test_key_types.py:88:13: E221 multiple spaces before operator
./ipatests/test_integration/test_key_types.py:90:1: E302 expected 2 blank lines, found 1
./ipatests/test_integration/test_key_types.py:92:13: E221 multiple spaces before operator
./ipatests/test_integration/test_key_types.py:94:1: E302 expected 2 blank lines, found 1
./ipatests/test_integration/test_key_types.py:96:13: E221 multiple spaces before operator
./ipatests/test_integration/test_key_types.py:98:1: E302 expected 2 blank lines, found 1
./ipatests/test_integration/test_key_types.py:100:13: E221 multiple spaces before operator
./ipatests/test_integration/test_key_types.py:102:1: E302 expected 2 blank lines, found 1
./ipatests/test_integration/test_key_types.py:104:13: E221 multiple spaces before operator
./ipatests/test_integration/test_key_types.py:106:1: E302 expected 2 blank lines, found 1
./ipatests/test_integration/test_key_types.py:108:13: E221 multiple spaces before operator
./ipatests/test_integration/test_key_types.py:110:1: E302 expected 2 blank lines, found 1
./ipatests/test_integration/test_key_types.py:112:13: E221 multiple spaces before operator
./ipatests/test_integration/test_key_types.py:114:1: E302 expected 2 blank lines, found 1
./ipatests/test_integration/test_key_types.py:116:13: E221 multiple spaces before operator
./ipatests/test_integration/test_key_types.py:118:1: E302 expected 2 blank lines, found 1
./ipatests/test_integration/test_key_types.py:120:13: E221 multiple spaces before operator
./ipatests/test_integration/test_key_types.py:122:1: E302 expected 2 blank lines, found 1
./ipatests/test_integration/test_key_types.py:124:13: E221 multiple spaces before operator
```

```
************* Module ipatests.test_integration.test_key_types
ipatests/test_integration/test_key_types.py:5: [W0611(unused-import), ] Unused import re)
```

jrisc committed:
ipa-getkeytab: fix bug in enctypes list count

When providing an explicit  list of key types to ipa-getkeytab, a bug in
length counting was causing two keys to be created with the final type
instead of just one.

Signed-off-by: Julien Rische <jrische@redhat.com>

jrisc committed:
ipa-getkeytab: pick salt based on credentials type

When ipa_string_to_enctypes() is used by ipa-getkeytab to get
identifiers of key/salt types, but the salt type is not provided,
"normal" was used by default.

In IPA, password-based keys should have a "special" salt type, and
randomized keys should have a "normal" one. This commit takes this
parameter into account to select the appropriate salt type.

Signed-off-by: Julien Rische <jrische@redhat.com>

jrisc committed:
kdb: warn on invalid kdc.conf supported_entypes

When generating randomized keys using kadmin, the list of key/salt types
is not fetched from the default password policy. It uses the
"supported_enctypes" setting from kdc.conf. The KDB API does not allow
to override this behavior at the moment.

This commit loads the value of kdc.conf's "supported_enctypes" at
initialization type, and compare it to the list inferred from the
"permitted_enctypes" setting. If at least one key/salt type is not
present in kdc.conf's "supported_enctypes", a warning is logged just
after the list of supported key types used by the plugin.

Signed-off-by: Julien Rische <jrische@redhat.com>

duzda commented on doc/designs/automated-keytab-rotation.md:
```diff
@@ -0,0 +1,333 @@
+The feature produces logs, when keytabs can not be read, rotated etc.
+The logs are available in the `ipactl.log` file. Available at `/var/log/ipactl.log`.
+
+The automated deletion produces logs elsewhere // TODO: Where? How?
```
I think all of this is a complete lie and will need to figure how logging here works and if... The code does call `logger.error`, `logger.debug` but it's unavailable I think.

menonsudhir commented:
@jrisc @duzda 

One of the observations regarding the failures with TestKeyTypeAes/TestKeyTypeAesSha2ShortAesSha1ShortRc4 enctypes.

- The tests related to **test_different_permitted_enctypes** are failing after ipactl restart and in some case the krb5 service is giving a traceback. Not sure if this is a bug we are hitting upon.

Failed to start krb5kdc Service
Traceback (most recent call last):
  File "/usr/lib/python3.14/site-packages/ipaserver/install/ipactl.py", line 404, in start_service
    svchandle.start(
    ~~~~~~~~~~~~~~~^
        capture_output=get_capture_output(svc, options.debug)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/usr/lib/python3.14/site-packages/ipaplatform/redhat/services.py", line 212, in start
    self.rotated_keys = check_and_rotate_keytabs(
                        ~~~~~~~~~~~~~~~~~~~~~~~~^
        self, self.api.env.host, self.api.env.realm
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/usr/lib/python3.14/site-packages/ipapython/krb5util.py", line 540, in check_and_rotate_keytabs
    _generate_keys(krb_principal, keytab_entries)
    ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.14/site-packages/ipapython/krb5util.py", line 318, in _generate_keys
    ipautil.run(
    ~~~~~~~~~~~^
        [paths.IPA_GETKEYTAB, "-p", principal, "-k", keytab_filepath],
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/usr/lib/python3.14/site-packages/ipapython/ipautil.py", line 607, in run
    raise CalledProcessError(
        p.returncode, arg_string, output_log, error_log
    )
ipapython.ipautil.CalledProcessError: CalledProcessError(Command ['/usr/sbin/ipa-getkeytab', '-p', 'ldap/master.ipa.test@IPA.TEST', '-k', '/etc/dirsrv/ds.keytab'] returned non-zero exit status 9: 'SASL Bind failed\n    Local error: SASL(-1): generic failure: GSSAPI Error: Unspecified GSS failure.  Minor code may provide more information (KDC has no support for encryption type)\nFailed to bind to server!\nRetrying with pre-4.0 keytab retrieval method...\nSASL Bind failed\n    Local error: SASL(-1): generic failure: GSSAPI Error: Unspecified GSS failure.  Minor code may provide more information (KDC has no support for encryption type)\nFailed to bind to server!\nFailed to get keytab\n')

Results:
- TestKeyTypeAes
[http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/d2ef1488-3424-11f1-8c43-fa163e030f13/report.html?sort=result ](url)
- TestKeyTypeAesSha2ShortAesSha1ShortRc4
[http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/e3c58530-3424-11f1-91b2-fa163e76b5bb/report.html?sort=result](url)

duzda committed:
Add krb5util keytab rotation and ipa-keytab-cleaner

Add krb5util helper to compare service keytabs against the
current permitted enctypes, and schedule removal of obsolete
KVNOs after krbMaxRenewableAge.

Also adds ipa-keytab-cleaner, which is a new tool,
that is used for the deletions. This will be later used
together with systemd services.

Specfile and autotools changes are also done.

Signed-off-by: David Hanina <dhanina@redhat.com>

duzda committed:
Add automated keytab rotation design

Describe automatic rotation of IPA service keytabs when permitted
encryption types change, for example after a RHEL 8 to RHEL 9 FIPS
upgrade or a crypto-policy update.

The document covers keytabs in scope, principal mapping for ipa-getkeytab,
when rotation runs at IPA start, and deferred KVNO removal via
ipa-keytab-cleaner and transient systemd timers. The description
explains which keytabs are rotated and how. Also includes a
test plan and debugging notes.

Signed-off-by: David Hanina <dhanina@redhat.com>

duzda committed:
Extend platform paths and services for keytab rotation

Add platform paths for the keytab cleaner config directory and wire
krb5kdc start/restart so rotation runs after the KDC is up and can
signal when keys were rotated.

Signed-off-by: David Hanina <dhanina@redhat.com>

duzda committed:
Integrate keytab rotation in ipactl and krbinstance

Start krb5kdc before other IPA services so keytab rotation can run
with a live KDC. When rotation updates keytabs, stop and restart all
services so every daemon loads the new keys.

Refactors ipactl start/restart.

Signed-off-by: David Hanina <dhanina@redhat.com>

duzda committed:
Ipactl better stack trace

Print traceback.format_exc() when ipactl fails to start a service, so
unexpected errors during start/restart are easier to diagnose.

Signed-off-by: David Hanina <dhanina@redhat.com>

duzda committed:
Skip keytab rotation during CA install, cert update, and upgrade

Automatic rotation during krb5kdc restart breaks several paths,
when rotation is unwanted and unexpected. This just builds on top
of other commits and enables the skip where required.

Update the design doc with call sites where rotation is skipped and
where it still runs.

Signed-off-by: David Hanina <dhanina@redhat.com>

duzda committed:
Add option to skip keytab rotation

Add skip_keytab_rotation to default.conf (default false) and ipalib
constants so administrators and install code can disable automatic
rotation while encryption types or master keys are being changed.

Pass skip_keytab_rotation through platform krb5kdc start/restart and
add --skip-keytab-rotation to ipactl.

Signed-off-by: David Hanina <dhanina@redhat.com>

duzda committed:
Enable skip_keytab_rotation in mkey and key_types tests

Integration tests that deliberately change permitted enctypes or
exercise master key upgrade must not trigger automatic keytab rotation
on ipactl start.

Set skip_keytab_rotation = True in /etc/ipa/default.conf on test
hosts after install in test_mkey_upgrade and test_key_types.

Signed-off-by: David Hanina <dhanina@redhat.com>

rjeffman commented:
I'm removing `needs review` as the PR needs to be rebased.
